File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(node, visited = None):
            if not visited:
                visited = set()

            if not node:
                return

            visited.add(node)

            temp = node.left
            node.left = node.right
            node.right = temp

            if node.left not in visited:
                dfs(node.left, visited)
            if node.right not in visited:
                dfs(node.right, visited)
        
        dfs(root)
        return root

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        int[] interval = new int[]{Integer.MIN_VALUE, Integer.MAX_VALUE};
        return dfs(root, interval);
    }
    private boolean dfs(TreeNode root, int[] interval) {
        if (root == null) {
            return true;
        }
        if (root.val <= interval[0] || root.val >= interval[1]) {
            return false;
        }
        return dfs(root.left, new int[]{interval[0], root.val})
        && dfs(root.right, new int[]{root.val, interval[1]});
    }
}
